Code/Framework/AzCore/AzCore/Jobs/Internal/JobTypeEraser.h

```diff
@@ -0,0 +1,232 @@
+
+        TypeErasedJob() = default;
+
+        template<typename Lambda>
+        TypeErasedJob(JobDescriptor const& desc, Lambda&& lambda) noexcept
+            : m_descriptor{ desc }
+        {
+            JobTypeEraser<Lambda> eraser;
+            m_invoker = eraser.ErasedInvoker();
+            m_relocator = eraser.ErasedRelocator();
+            m_destroyer = eraser.ErasedDestroyer();
+
+            // NOTE: This code is conservative in that extended alignment requirements result in a heap
+            // spill, even if the lambda could have occupied a portion of the inline buffer with a base
+            // pointer adjustment.
+            if constexpr (sizeof(Lambda) <= BufferSize && alignof(Lambda) <= alignof(max_align_t))
+            {
+                TypedRelocate(AZStd::forward<Lambda>(lambda), m_buffer);
+                m_lambda = m_buffer;
+            }
+            else
+            {
+                // Lambda has spilled to the heap (or requires extended alignment)
+                m_lambda = reinterpret_cast<char*>(azmalloc(sizeof(Lambda), alignof(Lambda)));
+                TypedRelocate(AZStd::forward<Lambda>(lambda), m_lambda);
+            }
+        }
+
+        TypeErasedJob(TypeErasedJob&& other) noexcept;
+
+        TypeErasedJob& operator=(TypeErasedJob&& other) noexcept;
+
+        ~TypeErasedJob();
+
+        void Link(TypeErasedJob& other);
+
+        // Indicates if this job is a root of the graph (with no dependencies)
+        bool IsRoot();
+
+        void Init() noexcept
+        {
+            m_dependencyCount = m_inboundLinkCount;
+        }
+
+        void Invoke()
+        {
+            m_invoker(m_lambda);
+        }
+
+        uint8_t GetPriorityNumber() const
+        {
+            return static_cast<uint8_t>(m_descriptor.priority);
+        }
+
+    private:
+        friend class CompiledJobGraph;
+        friend class JobWorker;
+
+        // This relocation avoids branches needed if the lambda type is unknown
+        template<typename Lambda>
+        void TypedRelocate(Lambda&& lambda, char* destination)
+        {
+            if constexpr (AZStd::is_trivially_move_constructible_v<Lambda>)
+            {
+                memcpy(destination, reinterpret_cast<char*>(&lambda), sizeof(Lambda));
+            }
+            else if constexpr (AZStd::is_move_constructible_v<Lambda>)
+            {
+                new (destination) Lambda{ AZStd::move(lambda) };
+            }
+            else if constexpr (AZStd::is_copy_constructible_v<Lambda>)
+            {
+                new (destination) Lambda{ lambda };
+            }
+            else
+            {
+                static_assert(
+                    AZStd::is_move_constructible_v<Lambda> || AZStd::is_copy_constructible_v<Lambda>,
+                    "Job lambdas must be either move or copy constructible. Please verify that all captured data is move or copy "
+                    "constructible.");
+            }
+        }
```

Comment:
There needs to be some kind of type check to make sure that a non-const rvalue reference to Lambda has been supplied in the scenarios where a move needs to occur.
Also the Lambda type needs to be wrapped in a [AZStd::remove_cvref_t](https://en.cppreference.com/w/cpp/types/remove_cvref) to make sure the raw type is being checked and not a reference type.

Reply:
I opted to delete the member function overload binding `Lambda &` instead to avoid the need for `remove_cvref_t` and also prevent binding lvalue refs to lambdas